wal-g pgsql: local backup fails if path to "data" on command line is relative instead of absolute

### Database name

postgresql

# Issue description

When run `wal-g backup-push path/to/data/directory`, en error is triggered when `path/to/data/directory` is a relative path.

### Describe your problem

with a correct wal-g.json file, run `wal-g --config wal-g.json backup_push ../pgsql/data`  <- point to the data directory with a relative path

expected result : succesful backup from local database

instead :
```
ERROR: 2022/11/16 15:40:51.242628 Data directory read from Postgres (../pgsql/data) is different than as parsed (/home/dbuser/pgsql/data).
panic: Data directory read from Postgres (../pgsql/data) is different than as parsed (/home/dbuser/pgsql/data).

goroutine 1 [running]:
log.(*Logger).Panicf(0xc000146d20?, {0x1a8bdbb?, 0x10?}, {0xc000a7fae0?, 0x203000?, 0x203000?})
        /home/user/go/src/log/log.go:258 +0x74
github.com/wal-g/wal-g/internal/databases/postgres.(*BackupHandler).HandleBackupPush(0xc000503860)
        /home/user/code/github.com/LeGEC/wal-g/internal/databases/postgres/backup_push_handler.go:331 +0x3fd
github.com/wal-g/wal-g/cmd/pg.glob..func4(0x2c1bd80?, {0xc00063bb60, 0x1, 0x3?})
        /home/user/code/github.com/LeGEC/wal-g/cmd/pg/backup_push.go:97 +0x732
...
```

--- a/internal/databases/postgres/backup_push_handler.go
+++ b/internal/databases/postgres/backup_push_handler.go
@@ -304,28 +304,35 @@ func (bh *BackupHandler) uploadBackup() internal.TarFileSets {
 // HandleBackupPush handles the backup being read from Postgres or filesystem and being pushed to the repository
 // TODO : unit tests
 func (bh *BackupHandler) HandleBackupPush() {
-	folder := bh.Workers.Uploader.UploadingFolder
-	baseBackupFolder := folder.GetSubFolder(bh.Arguments.backupsFolder)
-	tracelog.DebugLogger.Printf("Base backup folder: %s", baseBackupFolder)
-
 	bh.CurBackupInfo.StartTime = utility.TimeNowCrossPlatformUTC()
 
 	if bh.Arguments.pgDataDirectory == "" {
-		if bh.Arguments.forceIncremental {
-			tracelog.ErrorLogger.Println("Delta backup not available for remote backup.")
-			tracelog.ErrorLogger.Fatal("To run delta backup, supply [db_directory].")
-		}
-		// If no arg is parsed, try to run remote backup using pglogrepl's BASE_BACKUP functionality
-		tracelog.InfoLogger.Println("Running remote backup through Postgres connection.")
-		tracelog.InfoLogger.Println("Features like delta backup are disabled, there might be a performance impact.")
-		tracelog.InfoLogger.Println("To run with local backup functionalities, supply [db_directory].")
-		if bh.PgInfo.pgVersion < 110000 && !bh.Arguments.verifyPageChecksums {
-			tracelog.InfoLogger.Println("VerifyPageChecksums=false is only supported for streaming backup since PG11")
-			bh.Arguments.verifyPageChecksums = true
-		}
-		bh.createAndPushRemoteBackup()
-		return
+		bh.handleBackupPushRemote()
+	} else {
+		bh.handleBackupPushLocal()
 	}
+}
+
+func (bh *BackupHandler) handleBackupPushRemote() {
+	if bh.Arguments.forceIncremental {
+		tracelog.ErrorLogger.Println("Delta backup not available for remote backup.")
+		tracelog.ErrorLogger.Fatal("To run delta backup, supply [db_directory].")
+	}
+	// If no arg is parsed, try to run remote backup using pglogrepl's BASE_BACKUP functionality
+	tracelog.InfoLogger.Println("Running remote backup through Postgres connection.")
+	tracelog.InfoLogger.Println("Features like delta backup are disabled, there might be a performance impact.")
+	tracelog.InfoLogger.Println("To run with local backup functionalities, supply [db_directory].")
+	if bh.PgInfo.pgVersion < 110000 && !bh.Arguments.verifyPageChecksums {
+		tracelog.InfoLogger.Println("VerifyPageChecksums=false is only supported for streaming backup since PG11")
+		bh.Arguments.verifyPageChecksums = true
+	}
+	bh.createAndPushRemoteBackup()
+}
+
+func (bh *BackupHandler) handleBackupPushLocal() {
+	folder := bh.Workers.Uploader.UploadingFolder
+	baseBackupFolder := folder.GetSubFolder(bh.Arguments.backupsFolder)
+	tracelog.DebugLogger.Printf("Base backup folder: %s", baseBackupFolder)
 
 	if utility.ResolveSymlink(bh.Arguments.pgDataDirectory) != bh.PgInfo.PgDataDirectory {
 		tracelog.ErrorLogger.Panicf("Data directory read from Postgres (%s) is different than as parsed (%s).",
@@ -402,11 +409,12 @@ func NewBackupHandler(arguments BackupArguments) (bh *BackupHandler, err error) 
 
 	uploader, err := ConfigureWalUploader()
 	if err != nil {
-		return bh, err
+		return nil, err
 	}
 	pgInfo, err := getPgServerInfo()
 	if err != nil {
-		return bh, err
+		return nil, err
+	}
 	}
 
 	if arguments.pgDataDirectory != "" && arguments.pgDataDirectory != pgInfo.PgDataDirectory {
@@ -423,7 +431,7 @@ func NewBackupHandler(arguments BackupArguments) (bh *BackupHandler, err error) 
 		PgInfo: pgInfo,
 	}
 
-	return bh, err
+	return bh, nil
 }
 
 func (bh *BackupHandler) runRemoteBackup() *StreamingBaseBackup {

--- a/internal/databases/postgres/backup_push_handler.go
+++ b/internal/databases/postgres/backup_push_handler.go
@@ -330,14 +330,20 @@ func (bh *BackupHandler) handleBackupPushRemote() {
 }
 
 func (bh *BackupHandler) handleBackupPushLocal() {
+	{
+		// The 'data' path provided on the command line must point at the same directory as the one listed by the Postgresql server.
+		// If mismatched, this means we aren't connected to the correct server. This is a fatal error.
+		fromCli := bh.Arguments.pgDataDirectory
+		fromServer := bh.PgInfo.PgDataDirectory // that value is expected to already be absolute and "unsymlinked"
+		if utility.AbsResolveSymlink(fromCli) != fromServer {
+			tracelog.ErrorLogger.Fatalf("Data directory from command line '%s' is not the same as Postgres' one '%s'", fromCli, fromServer)
+		}
+	}
+
 	folder := bh.Workers.Uploader.UploadingFolder
 	baseBackupFolder := folder.GetSubFolder(bh.Arguments.backupsFolder)
 	tracelog.DebugLogger.Printf("Base backup folder: %s", baseBackupFolder)
 
-	if utility.ResolveSymlink(bh.Arguments.pgDataDirectory) != bh.PgInfo.PgDataDirectory {
-		tracelog.ErrorLogger.Panicf("Data directory read from Postgres (%s) is different than as parsed (%s).",
-			bh.Arguments.pgDataDirectory, bh.PgInfo.PgDataDirectory)
-	}
 	bh.checkPgVersionAndPgControl()
 
 	if bh.Arguments.isFullBackup {
@@ -414,13 +420,6 @@ func NewBackupHandler(arguments BackupArguments) (bh *BackupHandler, err error) 
 	pgInfo, err := getPgServerInfo()
 	if err != nil {
 		return nil, err
-	}
-	}
-
-	if arguments.pgDataDirectory != "" && arguments.pgDataDirectory != pgInfo.PgDataDirectory {
-		warning := fmt.Sprintf("Data directory for postgres '%s' is not equal to backup-push argument '%s'",
-			arguments.pgDataDirectory, pgInfo.PgDataDirectory)
-		tracelog.WarningLogger.Println(warning)
 	}
 
 	bh = &BackupHandler{

--- a/utility/utility.go
+++ b/utility/utility.go
@@ -137,6 +137,19 @@ func ResolveSymlink(path string) string {
 	return resolve
 }
 
+// AbsResolveSymlink first tries to turn 'path' to an absolute path, then calls ResolveSymlink.
+// If 'path' can't be turned to an absolute path, keep the relative path.
+func AbsResolveSymlink(path string) string {
+	abs, err := filepath.Abs(path)
+	if err != nil {
+		// keep a trace of the errors, but keep going with 'path'
+		tracelog.DebugLogger.Printf("could not convert path to absolute path: %s", err)
+		abs = path
+	}
+
+	return ResolveSymlink(abs)
+}
+
 func GetFileExtension(filePath string) string {
 	ext := path.Ext(filePath)
 	if ext != "" {

